File: src/main/java/corp/soyummy/userfavorites/UserFavoriteService.java
package corp.soyummy.userfavorites;

import org.springframework.stereotype.Service;
import lombok.RequiredArgsConstructor;
import org.bson.types.ObjectId;
import java.util.List;

import corp.soyummy.userfavorites.dto.UserFavoriteCreateDto;
import corp.soyummy.userfavorites.dto.UserFavoriteUpdateDto;
import corp.soyummy.errors.ResourceNotFoundException;

@Service
@RequiredArgsConstructor
public class UserFavoriteService {

    private final UserFavoriteRepository userFavoriteRepository;

    public List<UserFavorite> getAllFavorites() {
        return userFavoriteRepository.findAll();
    }

    public UserFavorite getFavoriteById(String id) {
        return userFavoriteRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Favorite not found with id: " + id));
    }

    public List<UserFavorite> getFavoritesByRecipeId(String recipeId) {
        ObjectId recipeIdObj = new ObjectId(recipeId);
        return userFavoriteRepository.findByRecipeId(recipeIdObj);
    }

    public List<UserFavorite> getFavoritesByUserId(String userId) {
        ObjectId userIdObj = new ObjectId(userId);
        return userFavoriteRepository.findByUserId(userIdObj);
    }

    public UserFavorite createFavorite(UserFavoriteCreateDto createDto) {
        UserFavorite favorite = new UserFavorite();
        favorite.setRecipeId(createDto.getRecipeId());
        favorite.setUserId(createDto.getUserId());
        return userFavoriteRepository.save(favorite);
    }

    public UserFavorite updateFavorite(String id, UserFavoriteUpdateDto updateDto) {
        UserFavorite existingFavorite = getFavoriteById(id);
        if (existingFavorite == null) {
            throw new ResourceNotFoundException("Favorite not found with id: " + id);
        }

        existingFavorite.setRecipeId(updateDto.getRecipeId());
        return userFavoriteRepository.save(existingFavorite);
    }

    public void deleteFavorite(String id) {
        if (!userFavoriteRepository.existsById(id)) {
            throw new ResourceNotFoundException("Favorite not found with id: " + id);
        }

        userFavoriteRepository.deleteById(id);
    }
}

File: src/main/kotlin/com/soyummy/userfavorites/UserFavoriteService.kt
package com.soyummy.userfavorites

import org.springframework.stereotype.Service
import com.soyummy.userfavorites.dto.UserFavoriteCreateDto
import com.soyummy.userfavorites.dto.UserFavoriteUpdateDto
import com.soyummy.errors.ResourceNotFoundException
import org.bson.types.ObjectId

@Service
class UserFavoriteService(private val userFavoriteRepository: UserFavoriteRepository) {
    fun getAllFavorites(): List<UserFavorite> = userFavoriteRepository.findAll()

    fun getFavoriteById(id: String): UserFavorite = userFavoriteRepository.findById(id)
            .orElseThrow { ResourceNotFoundException("Favorite not found with id: $id") }

    fun getFavoritesByRecipeId(recipeId: String): List<UserFavorite> =
        userFavoriteRepository.findByRecipeId(ObjectId(recipeId))

    fun getFavoritesByUserId(userId: String): List<UserFavorite> =
        userFavoriteRepository.findByUserId(ObjectId(userId))

    fun createFavorite(createDto: UserFavoriteCreateDto): UserFavorite {
        if (createDto.recipeId.isBlank() || createDto.userId.isBlank())
            throw IllegalArgumentException("Missing required fields when adding a new favorite")

        val favorite = UserFavorite(
            recipeId = createDto.recipeId,
            userId = createDto.userId
        )
        return userFavoriteRepository.save(favorite)
    }

    fun updateFavorite(id: String, updateDto: UserFavoriteUpdateDto): UserFavorite {
        if (updateDto.recipeId.isBlank())
            throw IllegalArgumentException("Missing required fields when update a favorite")

        val existingFavorite = getFavoriteById(id)
        val updatedFavorite = existingFavorite.copy(recipeId = updateDto.recipeId)

        return userFavoriteRepository.save(updatedFavorite)
    }

    fun deleteFavorite(id: String) {
        if (!userFavoriteRepository.existsById(id))
            throw ResourceNotFoundException("Favorite not found with id: $id")

        userFavoriteRepository.deleteById(id)
    }
}